File: 2492. minimum-score-of-a-path-between-two-cities.java
class Solution {
    int find(int[] root, int i) {
        if (root[i] == i)
            return i;
        return root[i] = find(root, root[i]);
    }

    public int minScore(int n, int[][] roads) {
        int[] root = new int[n + 1];
        for (int i = 0; i <= n; i++)
            root[i] = i;

        for (int[] r : roads)
            root[find(root, r[0])] = find(root, r[1]);

        int res = 10001;
        for (int[] r : roads)
            if (find(root, r[0]) == find(root, 1))
                res = Math.min(res, r[2]);

        return res;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        System.out.println(sol.minScore(
            4,
            new int[][]{
                {1, 2, 9},
                {2, 3, 6},
                {2, 4, 5},
                {1, 4, 7}
            }
        )); // Expected: 5

        System.out.println(sol.minScore(
            4,
            new int[][]{
                {1, 2, 2},
                {1, 3, 4},
                {3, 4, 7}
            }
        )); // Expected: 2
    }
}

File: 2492. minimum-score-of-a-path-between-two-cities.py
class Solution:
    def minScore(self, n: int, roads: list[list[int]]) -> int:
        root = list(range(n + 1))

        def find(i: int) -> int:
            root[i] = find(root[i]) if root[i] != i else i
            return root[i]

        for x, y, _ in roads:
            root[find(x)] = find(y)

        res, g1 = 10001, find(1)
        for x, _, d in roads:
            if find(x) == g1:
                res = min(res, d)

        return res


if __name__ == "__main__":
    sol = Solution()

    print(sol.minScore(
        4,
        [
            [1, 2, 9],
            [2, 3, 6],
            [2, 4, 5],
            [1, 4, 7]
        ]
    ))  # Expected: 5

    print(sol.minScore(
        4,
        [
            [1, 2, 2],
            [1, 3, 4],
            [3, 4, 7]
        ]
    ))  # Expected: 2